### PR TITLE
Refactor CrtResource

### PR DESCRIPTION
**Pull request by alexw91**
*Issue #, if available:* N/A

*Description of changes:*
Refactors the `CrtResource` class in preparation of adding Http Support. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/software/amazon/awssdk/crt/CrtResource.java [hunk 1]
```diff
@@ -42,12 +46,42 @@ public static int getAllocatedNativeResourceCount() {
...
+            throw new IllegalStateException("Can't acquire >1 Native Pointer");
+        }
+
+        if (_ptr == NULL) {
+            throw new IllegalStateException("Can't acquire NULL Pointer");
+        }
+
+        String lastValue = NATIVE_RESOURCES.put(_ptr, this.getClass().getCanonicalName());
+
+        if (lastValue != null) {
```
Comment (justinboswell) [inline src/main/java/software/amazon/awssdk/crt/CrtResource.java]: This isn't a valid check, it is entirely possible to be given the same address when requesting memory for an object.

File: src/main/java/software/amazon/awssdk/crt/CrtResource.java [hunk 2]
```diff
@@ -14,16 +14,20 @@
...
 import java.util.Collections;
+import java.util.LinkedList;
 import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * This wraps a native pointer to an AWS Common Runtime resource. It also ensures
  * that the first time a resource is referenced, the CRT will be loaded and bound.
  */
-public class CrtResource {
+public class CrtResource implements Closeable, AutoCloseable {
```
Comment (justinboswell) [inline src/main/java/software/amazon/awssdk/crt/CrtResource.java]: You only need AutoCloseable. it is undesirable to use both, and AutoCloseable supercedes Closeable.

File: src/main/java/software/amazon/awssdk/crt/CrtResource.java (hunk 1 above)
Comment (alexw91) [reply src/main/java/software/amazon/awssdk/crt/CrtResource.java]: Won't you run into use-after-free if you have two `CrtResource`'s with the same native pointer and you close one of them?

File: src/main/java/software/amazon/awssdk/crt/CrtResource.java (hunk 2 above)
Comment (alexw91) [reply src/main/java/software/amazon/awssdk/crt/CrtResource.java]: Fixed

File: src/main/java/software/amazon/awssdk/crt/CrtResource.java (hunk 1 above)
Comment (justinboswell) [reply src/main/java/software/amazon/awssdk/crt/CrtResource.java]: No you're right, this is my mistake. Basically, there can only be 1 resource at a time per address. As long as dead resources are cleared (which they are), all is well. I'm going to chalk that one up to pre-coffee code review :)

**Pull request merged**